File: src/lib/api/tags.ts
/**
 * Tag APIs
 *
 * API functions for Tag entity.
 */

import type { Tag, EntityType } from "@/lib/types";
import { withTauriError } from "./utils";

export async function getTag(id: string): Promise<Tag> {
  return withTauriError("get tag", async () => {
    const { invoke } = await import("@tauri-apps/api/core");
    return invoke<Tag>("get_tag", { id });
  });
}

export async function getTags(): Promise<Tag[]> {
  return withTauriError("get tags", async () => {
    const { invoke } = await import("@tauri-apps/api/core");
    return invoke<Tag[]>("get_tags");
  });
}

export async function createTag(
  name: string,
  color?: string,
  description?: string,
): Promise<Tag> {
  return withTauriError("create tag", async () => {
    const { invoke } = await import("@tauri-apps/api/core");
    return invoke<Tag>("create_tag", {
      name,
      color: color || null,
      description: description || null,
    });
  });
}

export async function updateTag(
  id: string,
  data: { name?: string; color?: string; description?: string },
): Promise<Tag> {
  return withTauriError("update tag", async () => {
    const { invoke } = await import("@tauri-apps/api/core");
    return invoke<Tag>("update_tag", {
      id,
      name: data.name || null,
      color: data.color || null,
      description: data.description || null,
    });
  });
}

export async function deleteTag(id: string): Promise<void> {
  return withTauriError("delete tag", async () => {
    const { invoke } = await import("@tauri-apps/api/core");
    return invoke<void>("delete_tag", { id });
  });
}

export async function getEntityTags(
  entityType: EntityType,
  entityId: string,
): Promise<Tag[]> {
  return withTauriError("get entity tags", async () => {
    const { invoke } = await import("@tauri-apps/api/core");
    return invoke<Tag[]>("get_entity_tags", { entityType, entityId });
  });
}

export async function setEntityTags(
  entityType: EntityType,
  entityId: string,
  tagIds: string[],
): Promise<void> {
  return withTauriError("set entity tags", async () => {
    const { invoke } = await import("@tauri-apps/api/core");
    return invoke<void>("set_entity_tags", { entityType, entityId, tagIds });
  });
}

export async function getEntitiesByTag(
  entityType: EntityType,
  tagIds: string[],
): Promise<string[]> {
  return withTauriError("get entities by tag", async () => {
    const { invoke } = await import("@tauri-apps/api/core");
    return invoke<string[]>("get_entities_by_tag", { entityType, tagIds });
  });
}

File: src-tauri/src/commands/tags.rs
// Tag CRUD commands

use super::validation::{normalize_color_or_default, validate_tag_name};
use crate::commands::repositories::tag_repository::{BulkTagResult, EntityTag, Tag};
use crate::commands::repositories::TagRepository;
use crate::log_command;
use crate::AppState;

#[tauri::command]
pub fn get_tags(state: tauri::State<AppState>) -> Result<Vec<Tag>, String> {
    log_command!("get_tags", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        TagRepository::get_all(&conn)
    })
}

#[tauri::command]
pub fn create_tag(
    state: tauri::State<AppState>,
    name: String,
    color: Option<String>,
    description: Option<String>,
) -> Result<Tag, String> {
    log_command!("create_tag", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;

        // Validate name using centralized validation
        validate_tag_name(&name)?;
        let name = name.trim().to_string();

        let id = uuid::Uuid::new_v4().to_string();

        // Validate and normalize color using centralized validation
        let color = normalize_color_or_default(&color.unwrap_or_default());

        TagRepository::create(&conn, &id, &name, &color, description.as_deref())
    })
}

#[tauri::command]
pub fn update_tag(
    state: tauri::State<AppState>,
    id: String,
    name: Option<String>,
    color: Option<String>,
    description: Option<String>,
) -> Result<Tag, String> {
    log_command!("update_tag", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;

        // Validate name if provided using centralized validation
        if let Some(ref n) = name {
            validate_tag_name(n)?;
        }

        // Validate and normalize color if provided
        let normalized_color = color.map(|c| normalize_color_or_default(&c));

        TagRepository::update(
            &conn,
            &id,
            name.as_deref(),
            normalized_color.as_deref(),
            description.as_deref(),
        )
    })
}

#[tauri::command]
pub fn delete_tag(state: tauri::State<AppState>, id: String) -> Result<(), String> {
    log_command!("delete_tag", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        TagRepository::delete(&conn, &id)
    })
}

#[tauri::command]
pub fn get_entity_tags(
    state: tauri::State<AppState>,
    entity_type: String,
    entity_id: String,
) -> Result<Vec<Tag>, String> {
    log_command!("get_entity_tags", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        TagRepository::get_by_entity(&conn, &entity_type, &entity_id)
    })
}

#[tauri::command]
pub fn set_entity_tags(
    state: tauri::State<AppState>,
    entity_type: String,
    entity_id: String,
    tag_ids: Vec<String>,
) -> Result<(), String> {
    log_command!("set_entity_tags", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        TagRepository::set_entity_tags(&*conn, &entity_type, &entity_id, &tag_ids)
    })
}

#[tauri::command]
pub fn get_entities_by_tag(
    state: tauri::State<AppState>,
    entity_type: String,
    tag_ids: Vec<String>,
) -> Result<Vec<String>, String> {
    log_command!("get_entities_by_tag", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        TagRepository::get_entities_by_tag(&conn, &entity_type, &tag_ids)
    })
}

/// Result for bulk tag operations (imported from tag_repository)

#[tauri::command]
pub fn bulk_add_tags(
    state: tauri::State<AppState>,
    entity_type: String,
    entity_ids: Vec<String>,
    tag_id: String,
) -> Result<BulkTagResult, String> {
    log_command!("bulk_add_tags", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        TagRepository::bulk_add_tags(&conn, &entity_type, &entity_ids, &tag_id)
    })
}

#[tauri::command]
pub fn bulk_remove_tags(
    state: tauri::State<AppState>,
    entity_type: String,
    entity_ids: Vec<String>,
    tag_id: String,
) -> Result<BulkTagResult, String> {
    log_command!("bulk_remove_tags", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        TagRepository::bulk_remove_tags(&conn, &entity_type, &entity_ids, &tag_id)
    })
}
